### PR TITLE
add FastVLM

### PR DESCRIPTION
**Pull request by yicycyc**
## PR 新增模型支持：FastVLM

本 PR 新增 `apple/FastVLM-0.5B` 支持。复现对齐对象为模型仓库中的 HuggingFace remote-code 实现及原始 safetensors 权重，验证使用完整的 0.5B checkpoint。

## 主要改动

- 新增 `paddleformers/transformers/fastvlm/`
  - `FastVLMConfig`
  - `FastVLMModel`
  - `FastVLMForConditionalGeneration`
  - `FastVLMVisionModel`
  - `FastVLMProcessor`
  - `FastVLMImageProcessor`
  - Qwen2 文本 backbone、MobileCLIP/FastViT 视觉塔及 MLP projector
  - 图像占位 token 展开、图文特征合并、loss 和 greedy generation
- 注册 FastVLM 到：
  - `paddleformers.transformers`
  - `AutoConfig`
  - `AutoModelForConditionalGeneration`
  - `AutoProcessor`
  - `AutoImageProcessor`
- 权重加载：
  - 支持直接加载原始 HuggingFace safetensors 目录
  - 转换 Qwen2、MobileCLIP/FastViT 和 projector 权重，并校验 missing、unexpected 和 shape mismatch
  - 用户无需预先运行离线权重转换脚本
- 新增 `tests/transformers/fastvlm/test_modeling.py`
  - 覆盖多模态 forward/loss、backward、AutoModel 加载和图像预处理

## 前向对齐验证

模型：`apple/FastVLM-0.5B`

验证设置：

- Transformers 与 PaddleFormers 加载相同原始权重和相同图文输入
- 单张 NVIDIA A100 40GB
- FP32
- attention backend：两侧均使用 eager
- 输出 logits shape：`[1, 283, 151936]`

结果：

| 对比范围 | max abs diff | mean abs diff | p99 abs diff | cosine | Top-1 一致率 |
| --- | ---: | ---: | ---: | ---: | ---: |
| 完整图文序列 | `4.96e-2` | `8.96e-4` | `6.31e-3` | `0.99999990` | `100%` |
| 有效文本位置 | `1.94e-3` | `4.67e-5` | `4.38e-4` | `0.9999999995` | `100%` |

满足前向对齐标准。

## 生成对齐

使用相同权重和输入进行 greedy generation，关闭采样，两侧前 10 个新 token 完全一致。

结论：前 10 个 token 匹配率为 `10/10`，生成对齐通过。

## 训练验证

PaddleFormers 与 ms-swift 使用相同的预处理结果和样本顺序，在完整 `apple/FastVLM-0.5B` 上完成单卡 BF16 full-parameter SFT。

共同设置：

- 单张 NVIDIA A100 40GB
- `max_steps = 300`
- `per_device_train_batch_size = 1`
- `gradient_accumulation_steps = 1`
- `learning_rate = 1e-5`
- `lr_scheduler_type = constant`
- `warmup_steps = 0`
- `weight_decay = 0`
- gradient clipping：global norm `1.0`
- `seed = 42`
- shuffle 关闭
- attention eager
- full-parameter SFT

### 1、纯文本

使用 GSM8K 300 条固定样本做 300-step full-SFT。

| 指标 | ms-swift | PaddleFormers | Paddle - ms-swift |
| --- | ---: | ---: | ---: |
| final loss | `0.57808` | `0.59387` | `+0.01578` |
| 300-step mean loss | `0.48863` | `0.54707` | `+0.05845` |
| last 50 mean loss | `0.50698` | `0.59787` | `+0.09088` |
| loss correlation | — | — | `0.95631` |

gsm_loss_comparison

### 2、多模态

使用 PaddleFormers 多模态 SFT 数据格式准备 300 条固定图文样本，两侧使用完全相同的 token、label、图像张量和样本顺序。

| 指标 | ms-swift | PaddleFormers | Paddle - ms-swift |
| --- | ---: | ---: | ---: |
| final loss | `3.01865` | `3.04167` | `+0.02303` |
| 300-step mean loss | `1.94669` | `1.93208` | `-0.01461` |
| last 50 mean loss | `1.77980` | `1.77458` | `-0.00523` |
| loss correlation | — | — | `0.98178` |

loss_comparison

纯文本和多模态训练均跑满 300 step，两侧 loss 收敛趋势一致。

### CODE REVIEW

File: paddleformers/transformers/fastvlm/modeling.py [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+        self.model = FastVLMModel(config)
+        self.lm_head = LMHead(config)
+        self.criterion = CriterionLayer(config)
+        self.tie_weights()
+
+    @classmethod
+    def from_pretrained(cls, pretrained_model_name_or_path, *args, **kwargs):
+        is_hf_safetensors = (
+            isinstance(pretrained_model_name_or_path, str)
+            and os.path.isdir(pretrained_model_name_or_path)
```
Comment (risemeup1111) [inline paddleformers/transformers/fastvlm/modeling.py]: P1

这里把专用转换分支限定为本地目录，导致标准的 `AutoModelForConditionalGeneration.from_pretrained("apple/FastVLM-0.5B")` 仍走父类通用加载：AutoModel 虽能从官方 `config.json` 选中本类，但传入这里的仍是模型 ID。官方权重的视觉键以 `model.vision_tower.vision_tower.model.*` 开头，而本模型需要 `model.vision_tower.vision_model.*`；当前通用 AoA 仅覆盖文本层和 projector，无法映射这些视觉参数。请让 Hub 模型 ID 解析到同一专用转换流程，或补齐完整视觉 AoA 映射，并增加直接从官方模型 ID 加载的回归测试。

File: paddleformers/transformers/auto/processing.py [hunk 2]
```diff
@@ -61,6 +61,7 @@
         ("glm4v_moe", "Glm4vProcessor"),
         ("glm_ocr", "Glm46VProcessor"),
         ("paligemma2", "PaliGemmaProcessor"),
+        ("llava_qwen2", "FastVLMProcessor"),
```
Comment (risemeup1111) [inline paddleformers/transformers/auto/processing.py:64]: P1

这条注册目前仍无法加载目标 checkpoint 的处理器。`apple/FastVLM-0.5B` 没有 `preprocessor_config.json` 或 `processor_config.json`；AutoProcessor 按 `model_type` 选中 `FastVLMProcessor` 后，`ProcessorMixin` 会先调用 `AutoImageProcessor.from_pretrained`，而该入口在缺少预处理配置且不是 timm 模型时会直接报错，无法利用这里的映射构造默认图像处理器。请为 FastVLM 的缺省预处理参数提供可加载路径，并用官方模型 ID 覆盖 `AutoProcessor.from_pretrained`。

File: paddleformers/transformers/fastvlm/modeling.py [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+class FastVLMModel(Qwen2Model):
+    """Qwen2 decoder augmented with the FastVLM MobileCLIP vision tower."""
+
+    config_class = FastVLMConfig
+
+    def __init__(self, config: FastVLMConfig):
+        config.fuse_rms_norm = False
+        super().__init__(config)
+        for layer in self.layers:
+            layer.self_attn.qkv_proj = FastVLMQKVLinear(config)
```
Comment (risemeup1111) [inline paddleformers/transformers/fastvlm/modeling.py]: P1

这里无条件用全量 `nn.Linear` 覆盖了 Qwen2 已按 TP 切分的 `qkv_proj`。以官方配置的 14 个 Q 头、2 个 KV 头为例，TP=2 时 `Qwen2Attention` 会按本地 1 个 KV 头 reshape，期望末维总宽度为 576，但这个替换层仍输出全局宽度 1152，首个注意力层即发生形状不匹配。请在 TP 下保留/实现 column-parallel QKV（同时保持当前融合顺序），并补 2 卡前向或与单卡的数值一致性测试。

File: paddleformers/transformers/fastvlm/modeling.py [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+        }
+        if inputs_embeds is not None and past_key_values is None:
+            model_inputs["inputs_embeds"] = inputs_embeds
+            model_inputs["input_ids"] = None
+        return model_inputs
+
+    @classmethod
+    def _gen_aoa_config(cls, config):
+        aoa_config = Qwen2ForCausalLMDeprecated._gen_aoa_config(config)
+        aoa_config["aoa_statements"].extend(
```
Comment (risemeup1111) [inline paddleformers/transformers/fastvlm/modeling.py:579]: P1

`save_pretrained` 默认通过 flex checkpoint 反向使用 `_gen_aoa_config`，但这里仅增加了 projector，未覆盖 `model.vision_tower.vision_model.*` 与原始 HF 视觉键之间的映射。这样导出的 `model.safetensors` 无法完整保存可重载的视觉塔；同时本类又会仅凭同名文件把本地 checkpoint 当成原始 HF 格式，标准 save/reload round-trip 会遗漏或拒绝这些参数。请补齐双向视觉规则（或使用可可靠判别的 Paddle 格式），并增加保存后重载的全参数一致性测试。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
| --- | --- | --- | --- |
| 1 | [modeling.py:248](https://github.com/PaddlePaddle/PaddleFormers/pull/4864#discussion_r3755557323) | P1 | ✅ |
| 2 | [processing.py:64](https://github.com/PaddlePaddle/PaddleFormers/pull/4864#discussion_r3755559729) | P1 | ✅ |
| 3 | [modeling.py:211](https://github.com/PaddlePaddle/PaddleFormers/pull/4864#discussion_r3755562437) | P1 | ✅ |
| 4 | [modeling.py:475](https://github.com/PaddlePaddle/PaddleFormers/pull/4864#discussion_r3755564563) | P1 | ✅ |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/fastvlm/modeling.py (hunk 1 above)
Comment (yicycyc) [reply paddleformers/transformers/fastvlm/modeling.py]: 已在 ea505d47 修复：Hub 模型 ID 会先解析单文件/分片 safetensors，再进入 FastVLM 专用视觉与文本转换；新增 apple/FastVLM-0.5B 模型 ID 回归测试。

File: paddleformers/transformers/auto/processing.py (hunk 2 above)
Comment (yicycyc) [reply paddleformers/transformers/auto/processing.py:64]: 已在 ea505d47 修复：FastVLMProcessor 在官方仓库缺少预处理 JSON 时，根据 mm_vision_tower 构造默认图像处理器；回归测试直接调用 AutoProcessor.from_pretrained("apple/FastVLM-0.5B")。

File: paddleformers/transformers/fastvlm/modeling.py (hunk 3 above)
Comment (yicycyc) [reply paddleformers/transformers/fastvlm/modeling.py]: 已在 ea505d47 修复：单卡保留精度对齐所需的专用 QKV GEMM，TP>1 保留 Qwen2 ColumnParallelLinear；新增 fused QKV/FFN 的两 rank 切分及与单卡投影输出数值一致性测试。

File: paddleformers/transformers/fastvlm/modeling.py (hunk 4 above)
Comment (yicycyc) [reply paddleformers/transformers/fastvlm/modeling.py:579]: 已在 ea505d47 修复：默认 save_pretrained 使用可可靠识别的原生 Paddle checkpoint，并提供对应原生重载路径；新增保存后重载的全参数逐项严格一致测试。

File: paddleformers/transformers/fastvlm/modeling.py (hunk 1 above)
Comment (risemeup1111) [reply paddleformers/transformers/fastvlm/modeling.py]: 已复核 `ea505d47`：`from_pretrained` 现会先解析远端 safetensors，并在识别官方视觉键后进入 FastVLM 专用转换；官方仓库当前的单文件权重路径也已覆盖。该项已解决。

File: paddleformers/transformers/auto/processing.py (hunk 2 above)
Comment (risemeup1111) [reply paddleformers/transformers/auto/processing.py:64]: 已复核 `ea505d47`：缺少预处理配置时，`FastVLMProcessor._get_arguments_from_pretrained` 会依据 `mm_vision_tower` 构造默认图像处理器，`AutoProcessor` 的本地映射路径可到达该分支。该项已解决。

File: paddleformers/transformers/fastvlm/modeling.py (hunk 3 above)
Comment (risemeup1111) [reply paddleformers/transformers/fastvlm/modeling.py]: 已复核 `ea505d47`：TP>1 时保留 Qwen2 的列并行 QKV 层，转换映射按 rank 切分融合 QKV/FFN；官方 14Q/2KV、TP=2 的本地宽度与注意力 reshape 一致。该项已解决。

File: paddleformers/transformers/fastvlm/modeling.py (hunk 4 above)
Comment (risemeup1111) [reply paddleformers/transformers/fastvlm/modeling.py:579]: 已复核 `ea505d47`：默认保存已固定为原生 Paddle checkpoint，本地重载会严格校验 missing/unexpected keys；全参数逐项一致的 round-trip 测试覆盖了视觉塔。该项已解决。

**Review by risemeup1111 — APPROVED** [review]

**Review by risemeup1111 — APPROVED** [review]

**Review by risemeup1111 — APPROVED** [review]
